File: web/src/core/feature/product/domain/caseuse/CheckAProductExistenceCaseUse.ts
import type {ProductRepository} from "../repository/product.repository";
import type {Sale, SaleItem} from "../../../sale/domain/entity/Sale";
import {availableStock} from "../entity/Product";

/**
 * Soft-check de disponibilidad usando stock disponible (existence - reserved).
 * No descuenta stock; solo valida que hay cupo para el soft-hold.
 */
export class CheckAProductExistenceCaseUse {
    constructor(private readonly repository: ProductRepository) {}

    async execute(sale: Sale): Promise<void> {
        if (sale.currency.length === 0) {
            throw new Error("No se puede ejecutar una compra vacía");
        }

        await this.checkSaleListExistence(sale.products);
    }

    private async checkSaleListExistence(saleList: SaleItem[]): Promise<void> {
        for (const item of saleList) {
            const exists = await this.checkExistence(
                item.quantity,
                item.productId
            );

            if (!exists) {
                throw new Error(
                    `No hay disponibilidad en la tienda para el producto: ${item.productName}`
                );
            }
        }
    }

    private async checkExistence(
        quantityOfSale: number,
        productId: string
    ): Promise<boolean> {
        const product = await this.repository.getById(productId);

        if (product === null) {
            throw new Error("El producto no se encuentra disponible");
        }

        return quantityOfSale <= availableStock(product);
    }
}

File: app/src/main/java/com/elitec/alejotaller/feature/product/domain/caseUse/CheckAProductExistenceCaseUse.kt
package com.elitec.alejotaller.feature.product.domain.caseUse

import com.elitec.alejotaller.feature.product.domain.repository.ProductRepository
import com.elitec.shared.sale.feature.sale.domain.entity.Sale
import com.elitec.shared.sale.feature.sale.domain.entity.SaleItem

/**
 * Soft-check de disponibilidad usando available = existence - reserved.
 * No descuenta stock físico; aplica soft-hold por separado al registrar.
 */
class CheckAProductExistenceCaseUse(
    private val repository: ProductRepository
) {
    suspend operator fun invoke(sale: Sale): Result<Unit> = runCatching {
        if (sale.products.isEmpty()) {
            error("No se puede ejecutar una compra vacía")
        }
        checkSaleListExistence(sale.products)
    }

    private suspend fun checkSaleListExistence(saleList: List<SaleItem>) {
        for (item in saleList) {
            val available = checkExistence(item.quantity, item.productId)
            if (!available) {
                val label = item.productName?.takeIf { it.isNotBlank() } ?: item.productId
                error("No hay disponibilidad en la tienda para el producto: $label")
            }
        }
    }

    private suspend fun checkExistence(quantityOfSale: Int, productId: String): Boolean {
        val product = repository.getById(productId)
            ?: error("El producto no se encuentra disponible")
        return quantityOfSale <= product.availableStock()
    }
}
